Update EditGravatar

v18 opened the pull request:
This is part 1 of ~~3~~ 4 related PRs.

1. EditGravatar general changes: https://github.com/Automattic/wp-calypso/pull/12595
2. Ask users to verify their email first: https://github.com/Automattic/wp-calypso/pull/12596
3. Allow users to drag and drop image: https://github.com/Automattic/wp-calypso/pull/12597
4. Analytics: https://github.com/Automattic/wp-calypso/pull/13134

## There are 2 changes included in this PR:
1. Adding a note that Gravatars are public
2. Removing the Gravatar link from the /me sidebar when the `me/edit-gravatar` feature flag is enabled

Here's the behaviour of the note:

![info-pop-over](https://cloud.githubusercontent.com/assets/11487924/24420158/e1519972-13bf-11e7-9fa6-2b5938b4f226.gif)

## Testing instructions:
1. Check out this branch and `make run` as usual
2. Go to calypso.localhost:3000/me
3. See that clicking the Gravatar in the sidebar takes you to https://secure.gravatar.com/site/wpcom?wpcc-no-close
4. See that there is no EditGravatar component shown:
<img width="1048" alt="screen shot 2017-03-28 at 12 58 19 pm" src="https://cloud.githubusercontent.com/assets/11487924/24417417/4c09fb88-13b6-11e7-86dd-43fea63f511a.png">

5. Turn on the feature flag using `export ENABLE_FEATURES=me/edit-gravatar && make run`
6. Go to calypso.localhost:3000/me
7. See that clicking the Gravatar in the sidebar does not do anything
8. See the EditGravatar component:
![edit-gravatar-component](https://cloud.githubusercontent.com/assets/11487924/24436632/9b06aac0-140a-11e7-92fc-0afc70c9c3a5.gif)

9. See the extra info about Gravatars, and can click the popover

## The wording

The info popover wording could use a lookover, so I added the Needs Copy Review tag :) Here's what it says:

> The avatar you use on WordPress.com comes from Gravatar - a universal avatar service.
Your photo may be displayed on other sites where you use your email address rockstar@email.com.

blowery commented on client/blocks/edit-gravatar/index.jsx:
```diff
@@ -158,8 +159,10 @@ export class EditGravatar extends Component {
+			translate,
 			user
 		} = this.props;
+		const gravatarLink = `//gravatar.com/${ user.username || '' }`;
```
we should probably just force this to `https`. Gravatar supports it so there's no great reason to downgrade to http.

v18 replied:
sounds good, added in bbcbd31c286d94cde7ebfae3db74608ce4a00110

blowery commented on client/blocks/edit-gravatar/index.jsx:
```diff
@@ -187,6 +190,28 @@ export class EditGravatar extends Component {
+								components: {
+									a: <a
+										href={ gravatarLink }
+										target="_blank"
```
we usually flag links that open in a new window with an "external" icon. see https://github.com/Automattic/wp-calypso/tree/master/client/components/external-link

v18 replied:
👍 , updated in 8cfaaabba81ea4a2c44526cf9c605b397bc825fe

blowery commented on client/me/profile-gravatar/index.jsx:
```diff
@@ -25,6 +27,18 @@ export default React.createClass( {
+		if ( isEnabled( 'me/edit-gravatar' ) ) {
+			return (
+				<div className="profile-gravatar">
+					<Gravatar user={ this.props.user } size={ 150 } imgSize={ 400 } />
```
why is imgSize `400`? might add a comment detailing how that size was derived, or add the derivation as a `const`, like 
```
const GRAVATAR_SIZE = 150;
const GRAVATAR_IMAGE_SIZE = ( GRAVATAR_SIZE * 2 ) + 100; // add 100 because ...
```

v18 replied:
I'm not 100% clear on why 400 is used, in this case I'm continuing to use the same imgSize that was used before in the same place.

From a search through the codebase, it looks like we are using imgSize=400 for larger images (larger than the default 96) for caching purposes. Adding that in as a comment.

v18 replied:
Updated in 022e9bfebe0bc98b9640181ab74ca9c779564d17

blowery commented on client/me/profile-gravatar/index.jsx:
```diff
@@ -3,17 +3,19 @@
 const debug = debugFactory( 'calypso:me:sidebar-gravatar' );
 
-export default React.createClass( {
+const ProfileGravatar = React.createClass( {
```
any chance we can get off `createClass` while we're here?

v18 replied:
Agreed. The code will either be refactored in a future tracking PR for this feature, or will be removed entirely when the feature launches.

blowery commented:
Woot, changes all look good to me. :)

blowery commented:
btw, you can run with the feature on with just `ENABLE_FEATURES=me/edit-gravatar make run`

blowery commented on client/blocks/edit-gravatar/index.jsx:
```diff
@@ -187,6 +191,29 @@ export class EditGravatar extends Component {
+						{ translate( '{{p}}The avatar you use on WordPress.com comes ' +
+							'from {{ExternalLink}}Gravatar{{/ExternalLink}} - a universal avatar service.{{/p}}' +
+							'{{p}}Your photo may be displayed on other sites where ' +
+							'you use your email address rockstar@email.com.{{/p}}',
```
did you mean to include the current email here instead of `rockstar@emailcom`?

v18 replied:
I sure did! 😱 updated in 42df165e57602d04ef3aad73ff7853e0f9b156fa

blowery commented:
tested out as expected, assuming it's expected that the edit grav component doesn't actually support editing the grav. :)

v18 commented:
@blowery thanks for the review! As for the `EditGravatar` component - it should be working, but does require some set-up. I didn't add the instructions in this PR since the `EditGravatar` functionality wasn't being changed, but if you'd like to test it out, here they are:

1. Find your WP.com bearer token for the user you are logged in as. You can do that using this script:
```
$ curl -# -X POST --data-urlencode "grant_type=password" --data-urlencode "client_id=appid" --data-urlencode "client_secret=appsecret" --data-urlencode "username=username" --data-urlencode "password=password" https://public-api.wordpress.com/oauth2/token
```
You can use Calypso's app ID & secret, found [here](https://github.com/Automattic/wp-calypso/blob/a1c610c300e040a2c101bb5af6a66f7cd72b52d6/config/production.json#L108-L109).
    If you have 2FA enabled, you can add this extra parameter to the call: `--data-urlencode     "wpcom_otp=one-time-password"`, or you can create a new application password to use instead.
2. Save the bearer token to localStorage using `localStorage.setItem('bearerToken', 'your_bearer_token');`
3. You can now use the feature in its entirety.

v18 commented:
Also wanted to surface what the new external link with the icon looks like:

<img width="310" alt="screen shot 2017-04-12 at 6 28 44 pm" src="https://cloud.githubusercontent.com/assets/11487924/24982187/2738af94-1fae-11e7-976e-0752fd50a822.png">

cc @drw158 or @iamtakashi for a design opinion on the visual change :)

gwwar commented:
Maybe set the icon size to 14x14 and reduce the left margin to 2px?

![screen shot 2017-04-12 at 5 13 51 pm](https://cloud.githubusercontent.com/assets/1270189/24984864/26d29ce4-1fa5-11e7-85b5-7419bbe43cf1.png)

gwwar commented on client/me/profile-gravatar/index.jsx:
```diff
@@ -23,32 +26,59 @@ export default React.createClass( {
+						href={ profileURL }
+						target="_blank"
+						rel="noopener noreferrer" >
+						@{ this.props.user.username }
```
This is unlikely to happen, but getCurrentUser can return null, so it's safer to do a check here:
```jsx
user && user.username
```

v18 replied:
Ooh yes. There are 3 of these, so I solved them all by falling back to an empty user object in 0f193e4f6a9de24bb4e5a9584f02fe5ff0b00d5d

v18 commented:
thanks @gwwar! I also added a pointer cursor, and updated the popover styling in , it looks like this now (in 3 browsers):

<img width="336" alt="screen shot 2017-04-13 at 12 24 41 am" src="https://cloud.githubusercontent.com/assets/11487924/24990194/15d2d306-1fe0-11e7-8612-9e9601f1dabc.png">
<img width="316" alt="screen shot 2017-04-13 at 12 24 35 am" src="https://cloud.githubusercontent.com/assets/11487924/24990192/15d0e438-1fe0-11e7-9310-544704bf7085.png">
<img width="302" alt="screen shot 2017-04-13 at 12 24 08 am" src="https://cloud.githubusercontent.com/assets/11487924/24990193/15d25232-1fe0-11e7-81ed-350747b7b288.png">


cc @drw158 @iamtakashi for the popover design change ^^ (new since earlier). Will rebase when ready :)

drw158 commented:
Looks pretty good. A couple of things:

1. The margin at the bottom looks pretty big! Not sure why.
2. We are really not suppose to use Gridicons smaller than 18px.

v18 commented:
@drw158 thanks! I updated it back to 18px, and removed the extra margin at the bottom. It now looks like this in Chrome:

 <img width="323" alt="screen shot 2017-04-15 at 12 57 41 am" src="https://cloud.githubusercontent.com/assets/11487924/25060903/a84acf6c-2176-11e7-8da6-6f72559bdcd0.png">

drw158 commented:
The margin at the bottom of the tooltip looks even bigger now :)

v18 commented:
Fixed the bottom margin issue, it now looks like this (@drw158-approved):

<img width="316" alt="screen shot 2017-04-17 at 12 43 46 pm" src="https://cloud.githubusercontent.com/assets/11487924/25096371/ea819eee-236d-11e7-9932-ea55bcf6c588.png">

v18 commented:
Rebased on master & updated the commits.

v18 commented:
@gwwar wanted to get your 👍 before merging. I updated `ProfileGravatar` to an ES6 class, since CircleCI was giving me an error for that due to the new linting error rule. The changes for that are in https://github.com/Automattic/wp-calypso/pull/12595/commits/549a8603dc03c8dc01a54912ce27eaba0e74deed

gwwar commented on client/me/profile-gravatar/index.jsx:
```diff
@@ -1,30 +1,55 @@
+		this.props.recordGoogleEvent( 'Me',
+			'Clicked on Unclickable Gravatar Image in Sidebar'
+		);
+	}
```
Don't forget the `;` for assignments

v18 replied:
Fixed, rebased, and pushed.

gwwar commented:
> I updated ProfileGravatar to an ES6 class, since CircleCI was giving me an error for that due to the new linting error rule.

Yeah not sure I'm a fan of the new strictness, as it tends to bloat diffs. Changes look good 👍 . Left a minor note.
